src/stream/src/executor/top_n/group_top_n_eowc.rs

```diff
@@ -0,0 +1,419 @@
+        self.state_table.init_epoch(epoch);
+        Ok(())
+    }
+}
+
+impl<S: StateStore, const WITH_TIES: bool> EowcGroupTopNExecutor<S, WITH_TIES> {
+    /// We remark that topN executor diffs from aggregate executor as it must output diffs
+    /// whenever it applies a batch of input data. Therefore, topN executor flushes data only
+    /// instead of computing diffs and flushing when receiving a barrier.
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    pub(crate) async fn top_n_executor_execute(self: Box<Self>) {
+        let input = self.input;
+        let mut inner = self.inner;
+
+        let mut input = input.execute();
+        let barrier = expect_first_barrier(&mut input).await?;
+        inner.init(barrier.epoch).await?;
+
+        yield Message::Barrier(barrier);
+
+        #[for_await]
+        for msg in input {
+            let msg = msg?;
+            match msg {
+                Message::Watermark(watermark) => {
+                    if watermark.col_idx == inner.group_by[0] {
+                        inner.window_watermark = Some(watermark.with_idx(0));
+                    }
+                }
+                Message::Chunk(chunk) => {
+                    inner.group_keys.apply_chunk(&chunk).await?;
+                    inner.state_table.write_chunk(chunk);
+                }
+                Message::Barrier(barrier) => {
+                    #[for_await]
+                    for chunk in inner.flush_data(barrier.epoch) {
+                        yield Message::Chunk(chunk?);
+                    }
+
+                    if let Some(watermark) = &inner.window_watermark {
+                        yield Message::Watermark(watermark.clone());
+                    }
+
+                    // Update the vnode bitmap, only used by Group Top-N.
+                    if let Some(vnode_bitmap) = barrier.as_update_vnode_bitmap(self.ctx.id) {
+                        let (_previous_vnode_bitmap, _cache_may_stale) =
+                            inner.state_table.update_vnode_bitmap(vnode_bitmap);
+                    }
+
+                    yield Message::Barrier(barrier)
+                }
+            };
+        }
+    }
+}
+
+impl<S: StateStore, const WITH_TIES: bool> Executor for EowcGroupTopNExecutor<S, WITH_TIES> {
+    fn execute(self: Box<Self>) -> crate::executor::BoxedMessageStream {
+        self.top_n_executor_execute().boxed()
+    }
+
+    fn schema(&self) -> &Schema {
+        &self.inner.info.schema
+    }
+
+    fn pk_indices(&self) -> PkIndicesRef<'_> {
+        &self.inner.info.pk_indices
+    }
+
+    fn identity(&self) -> &str {
+        &self.inner.info.identity
+    }
+
+    fn info(&self) -> ExecutorInfo {
+        self.inner.info.clone()
+    }
+}
+
+/// This is similar to `ColumnDeduplicater` for distinct aggregation, but simpler.
+/// It is for getting all the group keys, while `ColumnDeduplicater` deduplicates the
+/// distinct columns in each group.
```

Comment:
Can we merge them as a common component? I think they are both to do distinct by some key. For `ColumnDeduplicater` the distinct key is `group key + distinct columns` and here the key is `group key`

Reply:
Updated. I thought it would be dirty but it turns out not too hard. This may sacrifice some performance and readability, although maybe also improve some maintainability. Not sure if it's worth it.

```rust
// `Box<[i64]>` not needed for TopN
type DedupCache = ManagedLruCache<CompactedRow, Box<[i64]>>;

// Only one vis and `mut` not needed
mut visibilities: Vec<&mut Vis>,


// This is changed and not very friendly for both. Introduced additional vis check
let chunk = DataChunk::new(distinct_columns.to_vec(), n_rows);
...
for (datum_idx, (op, row)) in ops.iter().zip_eq_debug(chunk.rows()).enumerate() {
```